File: languages/ruby/Loggy/lib/loggy/logger.rb
require_relative "observable_logger"
require_relative "time_stamped_log_message"
require 'time'

module Loggy
    class Logger < ObservableLogger
        def log(log_message)
            time_stamped_log_message = TimeStampedLogMessage.new(log_message)
            @logger_observers.each do |logger_observer|
                logger_observer.process_log_message(time_stamped_log_message)
            end
        end
    end
end

File: code/packages/csharp/loggy-library/Loggy/Logger.cs
using System;
using System.Collections;
using System.Globalization;

namespace Loggy
{
    public class Logger : ILogger
    {
        private ArrayList loggerObservers;

        public Logger()
        {
            loggerObservers = new ArrayList();
        }

        public void Log(string logMessage)
        {
            TimeStampedLogMessage timeStampedLogMessage = new TimeStampedLogMessage();
            timeStampedLogMessage.logMessage = logMessage;
            timeStampedLogMessage.timeStamp = DateTime.UtcNow.ToString("yyyyMMddHHmmssffff");
            foreach(ILoggerObserver loggerObserver in loggerObservers)
            {
                loggerObserver.processLogMessage(timeStampedLogMessage);
            }
        }

        public void subscribe(ILoggerObserver loggerObserver)
        {
            loggerObservers.Add(loggerObserver);
        }
    }
}
